pkg/capabilities/consensus/ocr3/reporting_plugin_batching.go

```diff
@@ -0,0 +1,297 @@
+	}
+	if mid >= len(o.reqMap) {
+		return o
+	}
+
+	weids := make([]string, 0, len(o.reqMap))
+	for k := range o.reqMap {
+		weids = append(weids, k)
+	}
+
+	weids = weids[:mid]
+
+	newMap := make(map[string]*requests.Request, len(weids))
+	for _, weid := range weids {
+		if req, ok := o.reqMap[weid]; ok {
+			newMap[weid] = req
+		}
+	}
+
+	return ObservationSerializable{
+		reqMap: newMap,
+	}
+}
+
+type OutcomeSerializable struct {
+	previousOutcome *pbtypes.Outcome
+	weids           []*pbtypes.Id
+	outcomes        map[string]*pbtypes.AggregationOutcome
+}
+
+func (o OutcomeSerializable) Serialize(_ logger.Logger) ([]string, []byte, error) {
+	currentReports := make([]*pbtypes.Report, 0)
+	allExecutionIDs := make([]string, 0)
+
+	for _, weid := range o.weids {
+		outcome := o.outcomes[weid.WorkflowExecutionId]
+		report := &pbtypes.Report{
+			Outcome: outcome,
+			Id:      weid,
+		}
+		currentReports = append(currentReports, report)
+		allExecutionIDs = append(allExecutionIDs, weid.WorkflowExecutionId)
+
+		o.previousOutcome.Outcomes[weid.WorkflowId] = outcome
+	}
+
+	o.previousOutcome.CurrentReports = currentReports
+	rawOutcome, err := proto.MarshalOptions{Deterministic: true}.Marshal(o.previousOutcome)
+
+	return allExecutionIDs, rawOutcome, err
+}
+
+func (o OutcomeSerializable) Len() int {
+	return len(o.weids)
+}
+
+func (o OutcomeSerializable) Mid(mid int) Serializable {
+	if mid < 0 {
+		return OutcomeSerializable{
+			o.previousOutcome,
+			[]*pbtypes.Id{},
+			make(map[string]*pbtypes.AggregationOutcome),
+		}
+	}
+	if mid >= len(o.outcomes) {
+		return o
+	}
+
+	weids := make([]*pbtypes.Id, 0, len(o.weids))
+	for _, k := range o.weids {
+		weids = append(weids, k)
+	}
+
+	weids = weids[:mid]
+
+	newOutcomes := make(map[string]*pbtypes.AggregationOutcome, len(weids))
+	for _, weid := range weids {
+		if outcome, ok := o.outcomes[weid.WorkflowExecutionId]; ok {
+			newOutcomes[weid.WorkflowExecutionId] = outcome
+		}
+	}
+
+	return OutcomeSerializable{
+		o.previousOutcome,
+		weids,
+		newOutcomes,
+	}
+}
+
+// packToSizeLimit function maximizes the number of requests being included in a batch.
+// It finds the best utilization of space with the protobuf-marshalled structures using logarithmic (binary search)
+// approach to identify the optimal number of Requests that can be serialized without exceeding
+// the limit (defaultBatchSizeMiB).
+func packToSizeLimit(lggr logger.Logger, all Serializable) ([]string, []byte, error) {
+	var best []byte
+	var bestRequests Serializable
+	var bestExecutionIDs []string
+
+	if all.Len() == 0 {
+		// _, empty, _ := all.Serialize(lggr)
+		// return bestExecutionIDs, empty, nil
+		return nil, nil, errors.New("no requests to pack")
+	}
+
+	low, high := 0, all.Len()
+
+	optRound := 0
+
+	for low < high {
+		mid := (low + high) / 2
+		if mid == 0 {
+			mid = 1 // poor man's ceil
+		}
+		candidate := all.Mid(mid)
+		executionIDs, serialized, err := candidate.Serialize(lggr)
```

Comment:
We should probably be extra careful here in case the leader goes crazy and sends a very large query. Maybe a maxmax const for protection?

I also realized that we don't have deduplication of request IDs inside query - something we could add.